src/heuristics/Doc2Vec.py

```diff
@@ -0,0 +1,68 @@
+import string
+
+import nltk
+import numpy as np
+from gensim.models.doc2vec import Doc2Vec
+from nltk.corpus import stopwords
+
+from heuristics.Heuristics import AbstractHeuristic
+
+
+class Doc2VecHeuristic(AbstractHeuristic):
+    def __init__(self, cleaned):
+        self.model = Doc2Vec.load("../models/" + ("doc2vec_cleaned" if cleaned else "doc2vec"))
+        self.api = None
+        self.goal_vector = None
+        self.summaries = {}
+        self.goal = None
+        self.cleaned = cleaned
+
+    def setup(self, api, start, goal):
+        self.api = api
+        goal_summary_array = self.get_summary_array(goal)
+        self.goal_vector = self.model.infer_vector(doc_words=goal_summary_array, alpha=0.025, steps=20)
+        self.goal = goal
+
+    def preprocess_neighbors(self, neighbors):
+        new_dict = self.get_multiple_summary_arrays(neighbors)
+        self.summaries.update(new_dict)
+
+    def calculate_heuristic(self, node):
+        node_summary_array = self.get_summary_array(node)
+        node_summary_vector = self.model.infer_vector(node_summary_array)
+        return 0 if node == self.goal else 1 - abs(self.cos_sim(node_summary_vector, self.goal_vector))
+
+    def get_summary_array(self, node):
+        node_summary = self.summaries[node] if node in self.summaries else nltk.tokenize.word_tokenize(self.api.get_summaries([node])[node])
+        if self.cleaned:
+            node_summary = self.clean(node_summary)
+        return node_summary
+
+    def get_multiple_summary_arrays(self, neighbors):
+        summaries = self.api.get_summaries(neighbors)
+        if self.cleaned:
+            new_dict = {}
+            for neighbor in summaries:
```

Comment:
You can use dictionary comprehension here if you want, something like:
```
new_dict = {neighbor: self.clean(nltk.tokenize.word_tokenize(summaries[neighbor])) for neighbor in summaries}
```

Reply:
Wow magic